### PR TITLE
Add python version flag to the pyfunc serialized format

### PR DESCRIPTION
**Pull request by dbczumar**
* Adds the version of python with which a model was serialized to `pyfunc`

**Test Plan:**

This PR augments existing model exports tests to validate the addition of the python version to the serialized model

### CODE REVIEW

**Review by tomasatdatabricks — COMMENTED** [review]
Looks good. I think the part where we activate the environment could be make a bit cleaner so that it happens only in one place.

Also, could you add tests? there is a utility function to score model in Sagemaker (tests.helper_functions.py/score_model_in_sagemaker_docker_container), you could use that to test this new functionality.

File: mlflow/sagemaker/container/__init__.py [hunk 1]
```diff
@@ -61,9 +69,13 @@ def _serve():
...
         raise Exception("Only supports pyfunc models and this is not one.")
     conf = m.flavors[pyfunc.FLAVOR_NAME]
+    model_py_version = None
+    if pyfunc.PY_VERSION in conf:
+        model_py_version = conf[pyfunc.PY_VERSION]
+
     bash_cmds = []
     if pyfunc.ENV in conf:
-        print("activating custom environment")
+        print_flush("activating custom Anaconda environment")
```
Comment (aarondav) [inline mlflow/sagemaker/container/__init__.py]: If this is a concern, should we consider setting PYTHONUNBUFFERED so all prints are flushed?

File: mlflow/pyfunc/__init__.py [hunk 2]
```diff
@@ -108,6 +110,7 @@ def add_to_model(model, loader_module, data=None, code=None, env=None):
     :return: Updated model configuration.
     """
     parms = {MAIN: loader_module}
+    parms[PY_VERSION] = PYTHON_VERSION
```
Comment (aarondav) [inline mlflow/pyfunc/__init__.py]: Could we add a unit test to confirm the python version is in fact embedded into the model?

File: mlflow/sagemaker/container/__init__.py [hunk 3]
```diff
@@ -61,9 +69,13 @@ def _serve():
     if pyfunc.FLAVOR_NAME not in m.flavors:
         raise Exception("Only supports pyfunc models and this is not one.")
     conf = m.flavors[pyfunc.FLAVOR_NAME]
+    model_py_version = None
+    if pyfunc.PY_VERSION in conf:
+        model_py_version = conf[pyfunc.PY_VERSION]
+
     bash_cmds = []
     if pyfunc.ENV in conf:
```
[the remark this answers is not in the document]
Comment (dbczumar) [reply mlflow/sagemaker/container/__init__.py]: 1. Refactored so that creation only occurs in one place. We obtain a `CondaEnv` object first; this object specifies the environment's name, configuration file (if applicable) and python version (if applicable).

2. I think we should be able to parse the configuration file yaml and check for the version. If it's not specified, we can use the model's version. I know this differs from our offline discussion - how do you feel about that approach?

File: mlflow/sagemaker/container/__init__.py [hunk 4]
```diff
@@ -72,8 +84,32 @@ def _serve():
...
-        os.system("conda env create -n custom_env -f {}".format(env_path_dst))
-        bash_cmds += ["source /miniconda/bin/activate custom_env"] + _server_dependencies_cmds()
+        env_name = "custom_env"
+        os.system("conda env create -n {en} -f {ep}".format(en=env_name, ep=env_path_dst))
+        bash_cmds += ["source /miniconda/bin/activate {en}".format(en=env_name)] + \
+                _server_dependencies_cmds()
+    elif model_py_version is None:
+        print_flush("The model does not specify a Python version or a custom Anaconda environment"
+                    " to use.")
+        _warn_potentially_incompatible_conda_env()
```
[the remark this answers is not in the document]
Comment (dbczumar) [reply mlflow/sagemaker/container/__init__.py]: Added a warning to the `load_pyfunc` method that can be suppressed.

File: mlflow/sagemaker/container/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dbczumar) [reply mlflow/sagemaker/container/__init__.py]: Implemented the configuration modification described in (2)!

File: mlflow/sagemaker/container/__init__.py [hunk 5]
```diff
@@ -97,6 +104,96 @@ def _serve():
...
+    try:
+        with open(env_path, "r") as f:
+            env_yaml = yaml.load(f)
+
+        deps = env_yaml["dependencies"]
+        env_py_versions = list(filter(
+            lambda item : "=" in item and item.split("=")[0].lower() == "python", deps))
+        env_contains_python = len(env_py_versions) > 0
+        if env_contains_python:
+            env_py_version = env_py_versions[-1].split("=")[1]
```
Comment (aarondav) [inline mlflow/sagemaker/container/__init__.py]: At this point in the code, it's possible `python` has no version specified, right? It may not be super useful, but I think it's valid to not include a version specification in a conda env.

I wonder if a conda had something like `python>3.5`, as well, I don't think we'd support that. Is there a crisp specification on what we can see in the dependency list? It's a bit hard to understand if the problem that this method is trying to solve is really tractable, or if we can only solve it in the most common case.

File: mlflow/sagemaker/container/__init__.py [hunk 6]
```diff
@@ -97,6 +104,96 @@ def _serve():
...
+
+
+def _add_py_version_to_conda_env_if_necessary(env_path, model_py_version):
+    try:
+        with open(env_path, "r") as f:
+            env_yaml = yaml.load(f)
+
+        deps = env_yaml["dependencies"]
+        env_py_versions = list(filter(
+            lambda item : "=" in item and item.split("=")[0].lower() == "python", deps))
```
Comment (aarondav) [inline mlflow/sagemaker/container/__init__.py]: Is it possible that a dependency of python causes python to be included at runtime resolution of the conda environment? E.g., `pip=10.0.1=py36_0`.

File: mlflow/sagemaker/container/__init__.py [hunk 7]
```diff
@@ -97,6 +104,96 @@ def _serve():
...
+        env_contains_python = len(env_py_versions) > 0
+        if env_contains_python:
+            env_py_version = env_py_versions[-1].split("=")[1]
+            print_flush("Using Python version specified by the custom Anaconda environment:" 
+                        " Python {epyv}.".format(epyv=env_py_version))
```
Comment (aarondav) [inline mlflow/sagemaker/container/__init__.py]: According to the documentation here:
https://conda.io/docs/user-guide/tasks/build-packages/package-spec.html

It seems spaces may exist instead of/in addition to equal signs.

Comment (aarondav) [reply mlflow/sagemaker/container/__init__.py]: Do we depend on `conda`? We may be able to leverage MatchSpec to help deal with these peculiarities:
https://github.com/conda/conda/blob/ce207022f257019cca55a79d025d9e324591bf0b/conda/models/match_spec.py#L72

File: mlflow/sagemaker/container/__init__.py [hunk 8]
```diff
@@ -97,6 +104,96 @@ def _serve():
...
+                            " Python {mpyv}".format(mpyv=model_py_version))
+                deps.append("python={mpyv}".format(mpyv=model_py_version))
+            else:
+                deps.append("python={dpyv}".format(dpyv=DEFAULT_CONDA_PYTHON_VERSION))
+                _warn_potentially_incompatible_conda_env()
+
+        with open(env_path, "w") as out:
+            yaml.dump(env_yaml, out, default_flow_style=False)
+    except Exception as e:
+        print(e)
```
Comment (aarondav) [inline mlflow/sagemaker/container/__init__.py]: What kind of exceptions do we expect? This may not in general be a great error (e.g., `str(e)` is empty if the message is empty). Could we just avoid catching errors and not have them thrown anywhere?

File: mlflow/sagemaker/container/__init__.py [hunk 9]
```diff
@@ -97,6 +104,96 @@ def _serve():
...
+    major_version = ".".join(py_version.split(".")[:2])
+    return (major_version in SUPPORTED_CONDA_MAJOR_PY_VERSIONS)
+
+
+def _warn_potentially_incompatible_conda_env():
+    """
+    Prints a warning message indicating that the default Anaconda environment,
+    which may not be compatible with the model, will be used for serving.
+    """
+    print_flush("Using the default Anaconda environment with Python {dpyv}, which may not be"
```
Comment (aarondav) [inline mlflow/sagemaker/container/__init__.py]: Maybe `[WARNING]` or something? I think there's a lot of logging messages expected by default, so might be good to point this one out.

File: mlflow/sagemaker/container/__init__.py [hunk 10]
```diff
@@ -97,6 +104,96 @@ def _serve():
...
+        print_flush("Activating custom Anaconda environment")
+        env = model_config[pyfunc.ENV]
+        env_path_src = os.path.join("/opt/ml/model", env)
+        _add_py_version_to_conda_env_if_necessary(env_path_src, model_py_version)
+        env_path_dst = os.path.join("/opt/mlflow/", env)
+        env_path_dst_dir = os.path.dirname(env_path_dst)
+        if not os.path.exists(env_path_dst_dir):
+            os.makedirs(env_path_dst_dir)
+        # TODO: should we test that the environment does not include any of the server dependencies?
+        # Those are gonna be reinstalled. should probably test this on the client side
```
Comment (aarondav) [inline mlflow/sagemaker/container/__init__.py]: What is server vs client in this context?

File: mlflow/sagemaker/container/__init__.py [hunk 11]
```diff
@@ -97,6 +104,96 @@ def _serve():
...
+                deps.append("python={dpyv}".format(dpyv=DEFAULT_CONDA_PYTHON_VERSION))
+                _warn_potentially_incompatible_conda_env()
+
+        with open(env_path, "w") as out:
+            yaml.dump(env_yaml, out, default_flow_style=False)
+    except Exception as e:
+        print(e)
+
+
+def _create_conda_env(env):
```
Comment (aarondav) [inline mlflow/sagemaker/container/__init__.py]: There's a lot of usage of the word `env`, it's hard to follow what we're expecting here. I do like that the factoring of `_get_conda_env` vs `_create_conda_env` seems useful for the sake of unit testing (easier to test the intermediate output), but I might propose that `_create_conda_env` invokes `_get_conda_env` and immediately desugar the response to be

`(config, py_version)`.

Why do we need `name`? Couldn't we just pick a uniform name for the conda env?

File: mlflow/sagemaker/container/__init__.py [hunk 12]
```diff
@@ -97,6 +104,96 @@ def _serve():
...
+    if env.config is not None:
+        cmd = "conda env create -n {en} -f {cfg}".format(en=env.name, cfg=env.config)
```
Comment (aarondav) [inline mlflow/sagemaker/container/__init__.py]: btw, it may be useful to note that file may be empty (but existent), and this would also work.

**Review by tomasatdatabricks — COMMENTED** [review]
I left a few comments. Mostly

1. I am not sure if changing user provided environment is a good idea. I would maybe leave that for later. 

2. I think the logic of deciding what to do with the env and python version could be simplified.

3. For the testing part, can we reduce the number of Docker tests by unit-testing the individual container functions directly? I think the Docker test can be reduced to just integration test.

**Comment by dbczumar** [comment]
After offline discussion with @aarondav and @tomasatdatabricks, we've concluded that it makes sense to develop a solution that enables an appropriate Anaconda environment to be inferred for a model at serialization time on a per-flavor basis. Accordingly, I've removed the sagemaker container modifications introduced by this PR.

Logging the python version with the `pyfunc` flavor is still useful for presenting incompatibility warnings to the user, so this functionality remains.

File: mlflow/pyfunc/__init__.py [hunk 13]
```diff
@@ -127,13 +136,27 @@ def load_pyfunc(path, run_id=None):
         raise Exception("Format '{format}' not found not in {path}.".format(format=FLAVOR_NAME,
                                                                             path=conf_path))
     conf = model.flavors[FLAVOR_NAME]
+    model_py_version = conf[PY_VERSION] if PY_VERSION in conf else None
```
Comment (aarondav) [inline mlflow/pyfunc/__init__.py]: nit: `conf.get(PY_VERSION)`

File: mlflow/pyfunc/__init__.py [hunk 14]
```diff
@@ -127,13 +136,27 @@ def load_pyfunc(path, run_id=None):
...
+    if not suppress_warnings:
+        _warn_potentially_incompatible_py_version_if_necessary(model_py_version=model_py_version)
     if CODE in conf and conf[CODE]:
         code_path = os.path.join(path, conf[CODE])
         sys.path = [code_path] + _get_code_dirs(code_path) + sys.path
     data_path = os.path.join(path, conf[DATA]) if (DATA in conf) else path
     return importlib.import_module(conf[MAIN]).load_pyfunc(data_path)
 
 
+def _warn_potentially_incompatible_py_version_if_necessary(model_py_version=None):
```
Comment (aarondav) [inline mlflow/pyfunc/__init__.py]: Don't need `=None` since only one caller

File: mlflow/pyfunc/__init__.py [hunk 15]
```diff
@@ -127,13 +136,27 @@ def load_pyfunc(path, run_id=None):
...
+    if model_py_version is None:
+        print("The specified model does not have a specified Python version. It may be incompatible"
+              " with the version of Python that is currently running: Python {spyv}".format(
```
Comment (aarondav) [inline mlflow/pyfunc/__init__.py]: Maybe `{version}`?

File: mlflow/pyfunc/__init__.py [hunk 16]
```diff
@@ -127,13 +136,27 @@ def load_pyfunc(path, run_id=None):
...
+                   spyv=PYTHON_VERSION))
+    elif model_py_version != PYTHON_VERSION:
+        print("The version of Python that the model was saved in, Python {mpyv}, differs from the"
+              " version of Python that is currently running, Python {spyv},"
+              " and may be incompatible.".format( mpyv=model_py_version, spyv=PYTHON_VERSION))
```
Comment (aarondav) [inline mlflow/pyfunc/__init__.py]: nit: no space. Maybe `{specified_version}` and `{model_version}`? A bit of verbosity never hurt anyone.

File: mlflow/pyfunc/__init__.py [hunk 17]
```diff
@@ -127,13 +136,27 @@ def load_pyfunc(path, run_id=None):
...
     data_path = os.path.join(path, conf[DATA]) if (DATA in conf) else path
     return importlib.import_module(conf[MAIN]).load_pyfunc(data_path)
 
 
+def _warn_potentially_incompatible_py_version_if_necessary(model_py_version=None):
+    if model_py_version is None:
+        print("The specified model does not have a specified Python version. It may be incompatible"
+              " with the version of Python that is currently running: Python {spyv}".format(
+                   spyv=PYTHON_VERSION))
+    elif model_py_version != PYTHON_VERSION:
```
Comment (aarondav) [inline mlflow/pyfunc/__init__.py]: Can we avoid erroring on micro version differences?

File: mlflow/pyfunc/__init__.py (hunk 13 above)
Comment (dbczumar) [reply mlflow/pyfunc/__init__.py]: Done!

File: mlflow/pyfunc/__init__.py (hunk 14 above)
Comment (dbczumar) [reply mlflow/pyfunc/__init__.py]: Removed `=None`

File: mlflow/pyfunc/__init__.py (hunk 15 above)
Comment (dbczumar) [reply mlflow/pyfunc/__init__.py]: Yep! Changed `{spyv}` >> `{version}`

File: mlflow/pyfunc/__init__.py (hunk 17 above)
Comment (dbczumar) [reply mlflow/pyfunc/__init__.py]: Added a function `mlflow.utils.get_major_minor_py_version`. We now call this before comparing the serialized and system Python versions.

File: mlflow/pyfunc/__init__.py (hunk 16 above)
Comment (dbczumar) [reply mlflow/pyfunc/__init__.py]: Done!

File: mlflow/pyfunc/__init__.py [hunk 18]
```diff
@@ -127,13 +136,28 @@ def load_pyfunc(path, run_id=None):
         raise Exception("Format '{format}' not found not in {path}.".format(format=FLAVOR_NAME,
                                                                             path=conf_path))
     conf = model.flavors[FLAVOR_NAME]
+    model_py_version = conf.get(PY_VERSION) if PY_VERSION in conf else None
```
Comment (aarondav) [inline mlflow/pyfunc/__init__.py]: `conf.get(PY_VERSION)` returns None if PY_VERSION not in conf, so we don't need the rest of the conditional.

File: mlflow/pyfunc/__init__.py [hunk 19]
```diff
@@ -127,13 +136,28 @@ def load_pyfunc(path, run_id=None):
...
     if CODE in conf and conf[CODE]:
         code_path = os.path.join(path, conf[CODE])
         sys.path = [code_path] + _get_code_dirs(code_path) + sys.path
     data_path = os.path.join(path, conf[DATA]) if (DATA in conf) else path
     return importlib.import_module(conf[MAIN]).load_pyfunc(data_path)
 
 
+def _warn_potentially_incompatible_py_version_if_necessary(model_py_version):
+    if model_py_version is None:
+        print("The specified model does not have a specified Python version. It may be incompatible"
```
Comment (aarondav) [inline mlflow/pyfunc/__init__.py]: These should use eprint, `from mlflow.utils.logging_utils import eprint`, to ensure we print to stderr instead of stdout.

File: mlflow/pyfunc/__init__.py (hunk 18 above)
Comment (dbczumar) [reply mlflow/pyfunc/__init__.py]: Fixed!

File: mlflow/pyfunc/__init__.py (hunk 19 above)
Comment (dbczumar) [reply mlflow/pyfunc/__init__.py]: Thanks, using `eprint` now!

File: mlflow/pyfunc/__init__.py [hunk 20]
```diff
@@ -127,13 +137,29 @@ def load_pyfunc(path, run_id=None):
...
 
+def _warn_potentially_incompatible_py_version_if_necessary(model_py_version):
+    if model_py_version is None:
+        eprint("The specified model does not have a specified Python version. It may be"
+              " incompatible with the version of Python that is currently running:"
+              " Python {version}".format(
+                   version=PYTHON_VERSION))
+    elif get_major_minor_py_version(model_py_version) != get_major_minor_py_version(PYTHON_VERSION):
+        eprint("The version of Python that the model was saved in, Python {model_version}, differs"
+              " from the version of Python that is currently running, Python {system_version},"
```
Comment (aarondav) [inline mlflow/pyfunc/__init__.py]: lint will complain about this and the next line being one space off; as well as the eprint change above.

File: mlflow/pyfunc/__init__.py [hunk 21]
```diff
@@ -127,13 +137,29 @@ def load_pyfunc(path, run_id=None):
         raise Exception("Format '{format}' not found not in {path}.".format(format=FLAVOR_NAME,
                                                                             path=conf_path))
     conf = model.flavors[FLAVOR_NAME]
+    model_py_version = conf.get(PY_VERSION)
+    if not suppress_warnings:
+        _warn_potentially_incompatible_py_version_if_necessary(model_py_version=model_py_version)
```
Comment (tomasatdatabricks) [inline mlflow/pyfunc/__init__.py]: nit: That's a long function name, I would maybe shorten it to something like check_py_version. But it's only a suggestion.

**Review by tomasatdatabricks — APPROVED** [review]
Looks good to me. LGTM
